[IMP] inventory: update stock_warehouses (#3588)
odoo/documentation: content/applications/inventory_and_mrp/inventory/routes/concepts/stock_warehouses.rst

===========================================================
Sell stock from multiple warehouses using virtual locations
===========================================================

While keeping stock (and selling inventory) from one warehouse might work for smaller companies,
bigger companies might need to keep stock in (and sell from) multiple warehouses in multiple

> locations.

Reviewer: Can you explain why a company would use virtual locations? Is there an advantage? Why not use physical locations?

Author: I created a `.. spoiler::` tag to dive into this!

Sometimes, products included in a single sales order might take stock from two (or more)
:guilabel:`Warehouses`. In Odoo, this can be done by using *Virtual Locations*.

.. note::
   To create *Virtual Locations* in :guilabel:`Warehouses`, :guilabel:`Storage Locations` will need
   to be enabled. To do so, go to :menuselection:`Inventory --> Configuration --> Settings`, scroll
   down to the :guilabel:`Warehouse` section, click the checkbox next to
   :guilabel:`Storage Locations`, and :guilabel:`Save` changes.

Create and configure a virtual parent location
==============================================

Before creating any virtual stock *Locations*, a new *Warehouse* will need to be created. This new
Warehouse will act as a "Virtual Warehouse," and will be the "Parent" Location of other physical
Warehouses.

Create a new warehouse
----------------------

To create a new Warehouse, go to :menuselection:`Inventory --> Configuration --> Warehouses`, and
click :guilabel:`Create`. From here, the :guilabel:`Warehouse` name can be changed, and other

> details about how to resupply the :guilabel:`Warehouse` can be configured, as well. *Routes* can

Author: I'm able to see this in 14.0 - it's under the `Resupply` field. But, I realized that this wasn't too clear, so I added some instructions and pointed out where everything is here.

also be set by clicking on the :guilabel:`Routes` smart button. Once the :guilabel:`Warehouse` is
configured, virtual :guilabel:`Locations` can be created.

> .. image:: stock_warehouses/stock-warehouses-create-warehouse.png

Author: Right again... updating this!

   :align: center
   :alt: The edit screen for creating a new warehouse.

Create a virtual parent location
--------------------------------

.. important::
   In order to take stock from multiple Warehouses to fulfill a sales order, there need to be at
   least **two** warehouses acting as "Child Locations" of the "Virtual Parent" Location Warehouse.

To create and edit Locations, go to :menuselection:`Inventory --> Configuration --> Locations`. All
:guilabel:`Locations` are listed here, including the stock Location of the virtual Warehouse that
was created. Click into the stock :guilabel:`Location` for the virtual Warehouse that was

> previously created. Then, under the :guilabel:`Additional Information` tab, change the

Reviewer: ```suggestion
previously created. Then, under the :guilabel:`Additional Information` section, change the
```

> :guilabel:`Location Type` from :guilabel:`Internal Location` to :guilabel:`View`.

Reviewer: ```suggestion
:guilabel:`Location Type` from :guilabel:`Internal Location` to :guilabel:`View`. :guilabel:`Save` the changes.
```

This identifies this Location as a "Virtual" one, used to create a hierarchical structure for
a Warehouse, aggregating its "Child" Locations.

.. note::
   Products can *not* be stored in this :guilabel:`Location Type`.

.. image:: stock_warehouses/stock-warehouses-location-types.png
   :align: center
   :alt: Warehouse location types in location creation screen.

Configure physical warehouse locations
--------------------------------------

Navigate back to the :guilabel:`Locations` overview (via the breadcrumbs), then click into the

> first physical :guilabel:`Warehouse` stock :guilabel:`Location`.

Reviewer: Need to clarify this a bit better and specify which warehouse. I think you're referring to the one you previously created. I have to clear search out to find the locations.

Author: You're right, good call. I added some clarification here.

Under :guilabel:`Parent Location`, select the virtual :guilabel:`Warehouse` from the drop-down
menu. Then, navigate back to the :guilabel:`Locations` overview, and repeat this step for the
second physical :guilabel:`Warehouse` stock :guilabel:`Location`.

Both Locations are now "Child" Locations of the virtual warehouse "Parent" location.

> .. tip::

Reviewer: What's the advantage of doing this?

Author: I suppose that was a little ambiguous - I've added detail in to clarify.

   can have the Warehouse assigned to them from the drop-down menu next to
   :guilabel:`Default Warehouse` on their employee form.

.. image:: stock_warehouses/stock-warehouses-employee-form.png
   :align: center
   :alt: Default warehouse location on employee form.

Example flow: Sell a product from a virtual warehouse
=====================================================

To sell products from multiple Warehouses using a virtual "Parent" Location, there must be at least
*two* products and at least *two* Warehouses configured - with at least *one* product in each
Warehouse.

To create a new :abbr:`RFQ (Request for Quotation)`, navigate to the :guilabel:`Sales` app, and
click :guilabel:`Create`. Fill out the information on the new quotation by adding a
:guilabel:`Customer`, and click :guilabel:`Add a product` to add the two products stored in the two
Warehouses.

Then, click the :guilabel:`Other Info` tab, and change the :guilabel:`Warehouse` to the virtual
Warehouse that was previously created. Once the quotation has been filled out, :guilabel:`Confirm`.

Now that the quotation has been changed to a sales order, click the :guilabel:`Delivery` smart

> button. From here, change the :guilabel:`Source Location` to the virtual :guilabel:`Warehouse` that

Reviewer: If you create the order with a user that has the virtual warehouse assigned do we still need to make all these changes on the delivery order?

This seems like a lot of extra steps (almost a workaround and not a feature to reduce time)

Author: I showed this "long way 'round" method on purpose, to show that it can be done even if somebody is just doing it for the first time. I **did** write a `.. tip::`, however, saying that you can shorten this process by doing exactly what you suggested. I just moved that to the bottom of the page where it fits better!

warehouses.

.. spoiler:: Why a "virtual" warehouse?

   Virtual warehouses are great for companies with multiple physical warehouses. This is because
   a situation might arise when one warehouse runs out of stock of a particular product, but
   another warehouse still has stock on-hand. In this case, stock from these two (or more)
   warehouses could be used to fulfill a single sales order.

   The "virtual" warehouse acts as a single aggregator of all the inventory stored in a company's
   physical warehouses, and is used (for traceability purposes) to create a hierarchy of locations
   in Odoo.

Create a new warehouse
----------------------

To create a new warehouse, go to :menuselection:`Inventory --> Configuration --> Warehouses`, and
click :guilabel:`Create`. From here, the warehouse :guilabel:`Name` and :guilabel:`Short Name` can
be changed, and other warehouse details can be changed under the
:guilabel:`Warehouse Configuration` tab.

Under :guilabel:`Shipments`, the number of steps used to process :guilabel:`Incoming Shipments` and
:guilabel:`Outgoing Shipments` (1 step, 2 steps, or 3 steps) can be configured. Under

> :guilabel:`Resupply`, settings to resupply this warehouse can be configured. This includes whether

Reviewer 2: Wording
```suggestion
Under the :guilabel:`Resupply` heading, configure the method(s) for how the warehouse resupplies its
inventory:
```

> this warehouse should :guilabel:`Manufacture` products, :guilabel:`Buy to Resupply` products, or
> :guilabel:`Resupply Subcontractors`.

Reviewer 2: It's not helpful to the reader to reflect back surface-level information they can see for themselves. We can help guide them on their options by covering the form options, thoroughly.

Especially in Odoo 14, these options don't have any tooltips and the information on their use is hard to find on search.
```suggestion
- :guilabel:`Resupply Subcontractors`: resupply subcontractors with components from this warehouse.
- :guilabel:`Manufacture to Resupply`: when products are manufactured, they can be manufactured in
  this warehouse.
- :guilabel:`Manufacture`: to produce right away, move the components to the production location
  directly and start the manufacturing process; to pick first and then produce, unload the
  components from the stock to input location first, and then transfer it to the production
  location.
- :guilabel:`Buy to Resupply`: when products are bought, they can be delivered to this warehouse.
- :guilabel:`Resupply from`: automatically create routes to resupply this warehouse from another
  chosen warehouse
```

**two** products and at least **two** warehouses configured - with at least **one** product with
quantity on-hand in each warehouse, respectively.

To create a new Request for Quotation, or RFQ, navigate to the :menuselection:`Sales` app, and
click :guilabel:`Create` from the :guilabel:`Quotations` overview. Fill out the information on the
new quotation by adding a :guilabel:`Customer`, and click :guilabel:`Add a product` to add the two
products stored in the two warehouses.

Then, click the :guilabel:`Other Info` tab. Under the :guilabel:`Delivery` section, change the
:guilabel:`Warehouse` listed to the virtual warehouse that was previously created. Once the
warehouse has been changed, click :guilabel:`Confirm` to confirm the quotation and create a sales
order.

Now that the quotation has been confirmed as a sales order, click the :guilabel:`Delivery` smart
button. From the :guilabel:`Warehouse Delivery form`, click :guilabel:`Edit`, and change the
:guilabel:`Source Location` to the *Stock* location of the virtual warehouse that was previously
created, it it was not already automatically set as the *source location*.

> .. important::
>    The :guilabel:`Source Location` and :guilabel:`Warehouse`, under the :guilabel:`Other Info` tab
>    on the :guilabel:`Sales Order form`, must match in order for the products included in the sales
>    order to be pulled from different warehouses.

Reviewer 2: This is a very touchy setup if not performed correctly. 
Would include some basic troubleshooting for each possible error/scenario that might show up (as I ran into all 3 😓 )

```suggestion
.. important::
   The :guilabel:`Source Location` on the warehouse delivery form and the :guilabel:`Warehouse`
   under the :guilabel:`Other Info` tab on the sales order form *must* match in order for the
   products included in the sales order to be pulled from different warehouses.

  - If the virtual warehouse is not the value in the :guilabel:`Source Location` field on the
    warehouse delivery form, then click :guilabel:`Edit`, make the change, and click
    :guilabel:`Save`.
  - If the virtual warehouse is not the value in the :guilabel:`Warehouse` field on the sales order,
    then a new quotation may need to be generated.
  - If whe :guilabel:`Warehouse` field is missing on the sales order form, then the virtual
    warehouse (and its children warehouses) may not have been set up correctly, in which case,
    review the documentation above again to make sure all settings/configuration were done properly.
```

.. image:: stock_warehouses/stock-warehouses-delivery-order.png
   :align: center
   :alt: Delivery order with matching source and child locations.

> Finally, from the :guilabel:`Warehouse Delivery form`, if they aren't set already, click
> :guilabel:`Edit` and select the drop-down menus to change the *Locations* under the
> :guilabel:`From` column for each product to the *child locations* previously tied to the virtual
> *parent* location.

Reviewer 2: Paragraphs written this way are confusing to me.
- Where _exactly_ on the form?
- Give direct instructions, not "may or may not" hypotheticals 😉 

Note:
You may need to touch this section up later as the line items for this part of the delivery order did *not* automatically populate for me.

```suggestion
Finally, on the warehouse delivery form, under the :guilabel:`Detailed Operations` tab, confirm
that the *Locations* values under the :guilabel:`From` column for each product matches to the *child
locations* that are tied to the virtual *parent location*.
```

.. note::
   To view which *Locations* the products are coming from on the drop-down menus, click the

>    :guilabel:`Internal Link` button to expand the *Location* information. If needed, it can be
>    changed from here (granted there is quantity on-hand in that location).

Reviewer 2: ```suggestion
   :guilabel:`internal link (arrow)` icon to expand the *Location* information. If needed, it can be
   changed from here (granted there is quantity on hand for the product in that location).
```

> Once everything has been properly set and/or changed, click :guilabel:`Validate` then
> :guilabel:`Apply` to validate the delivery. Then, naviagte back to the :guilabel:`Sales Order form`
> (via the breadcrumbs), and click :guilabel:`Create Invoice` to invoice for the sales order.

Reviewer 2: - redundancy
- missing conjunction
- typo
- unnecessary guilabel

```suggestion
Once everything has been properly set, click :guilabel:`Validate` and then :guilabel:`Apply` to
validate the delivery. Then, navigate back to the sales order form (via the breadcrumbs), and
click :guilabel:`Create Invoice` to invoice for the sales order.
```

bigger companies might need to keep stock in and sell from multiple warehouses in multiple
locations.

Sometimes, products included in a single sales order might take stock from two or more warehouses.
In Odoo, this can be done by using **Virtual Locations**.

.. note::
   To create virtual locations in warehouses, :guilabel:`Storage Locations` will need to be
   enabled. To do so, go to :menuselection:`Inventory --> Configuration --> Settings`, scroll down
   to the **Warehouse** section, click the checkbox next to :guilabel:`Storage Locations`, and
   :guilabel:`Save` changes.

Create and configure a virtual parent location
==============================================

Before creating any virtual stock locations, a new warehouse will need to be created that will act
as a virtual warehouse, and will be the parent location of other physical warehouses.

Create a new warehouse
----------------------

To create a new warehouse, go to :menuselection:`Inventory --> Configuration --> Warehouses`, and
click :guilabel:`New`. From here, the warehouse name can be changed, as well as other details about

> how to resupply the warehouse. Routes can also be set by clicking on the :guilabel:`Routes` smart

Reviewer 3: From here, the warehouse name can be changed, and other details about
how to resupply the warehouse can be configured, as well.

button. Once the warehouse is configured, virtual locations can be created.

.. image:: stock_warehouses/stock-warehouses-create-warehouse.png
   :align: center
   :alt: The edit screen for creating a new warehouse.

Create a virtual parent location
--------------------------------

.. important::
   In order to take stock from multiple warehouses to fulfill a sales order, there need to be at

>    least **two** warehouses acting as children locations of the virtual parent location warehouse.

Reviewer 3: In order to take stock from multiple warehouses to fulfill a sales order, there must be at least **two** warehouses acting as "Children Locations" of the "Virtual Parent" location warehouse.

To create and edit locations, go to :menuselection:`Inventory --> Configuration --> Locations`. All
locations are listed here,including the stock location of the virtual warehouse that was created.
Click into the stock location for the virtual warehouse created previously, and under the
**Additional Information** tab, change the :guilabel:`Location Type` from *Internal Location* to
*View*. This identifies this location as a virtual one used to create a hierarchical structure for
a warehouse, aggregating its child locations. Products can *not* be stored in this

> :guilabel:`Location Type`.

Reviewer 3: Click into the stock location for the virtual warehouse that was previously created. Then, under the
::guilabel::`Additional Information` tab, change the :guilabel:`Location Type` from *Internal Location* to
*View*. 

This identifies this location as a "Virtual" one, used to create a hierarchical structure for
a warehouse, aggregating its "Child" locations. 

.. note::
   Products can *not* be stored in this :guilabel:`Location Type`.

.. image:: stock_warehouses/stock-warehouses-location-types.png
   :align: center
   :alt: Warehouse location types in location creation screen.

Configure physical warehouse locations

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer 3: Should this be "------" instead of "~~~~~". Double check the "Header Hierarchy" to see what's the more appropriate header format

Navigate back to the :guilabel:`Locations` overview via the breadcrumbs, then click into the first
physical warehouse stock location. Under :guilabel:`Parent Location`, select the virtual warehouse
from the drop-down menu. Navigate back to the :guilabel:`Locations` overview and repeat this step
for the second physical warehouse stock location. Both locations are now child locations of the

> virtual warehouse parent location.

Reviewer 3: Navigate back to the :guilabel:`Locations` overview (via the breadcrumbs), then click into the first
physical warehouse stock location. 

Under :guilabel:`Parent Location`, select the virtual warehouse
from the drop-down menu. Then, navigate back to the :guilabel:`Locations` overview, and repeat this step
for the second physical warehouse stock location. 

Both locations are now "Child" locations of the virtual warehouse "Parent" location.

.. tip::
   To use a virtual parent location as the default warehouse for sales orders, each salesperson can
   have the warheouse assigned to them from the drop-down menu next to
   :guilabel:`Default Warehouse` on their employee form.

.. image:: stock_warehouses/stock-warehouses-employee-form.png
   :align: center
   :alt: Default warehouse location on employee form.

Example flow: Sell a product from a virtual warehouse
=====================================================

To sell products from multiple warheouses using a virtual parent location, have at least *two*
products and at least *two* warehouses configured, with at least one product in each warehouse.

To create a new :abbr:`RFQ (request for quotation)`, navigate to the :guilabel:`Sales` app, and
click :guilabel:`New`. Fill out the inforamtion on the new quotation, and add the two products
stored in the two warehouses. Then, click the :guilabel:`Other Info` tab, and change the
:guilabel:`Warehouse` to the virtual warehouse created previously. Once the quotation has been

> filled out, :guilabel:`Confirm`.

Reviewer 3: To sell products from multiple warheouses using a virtual parent location, there must be at least *two*
products and at least *two* warehouses configured - with at least *one* product in each warehouse.

To create a new :abbr:`RFQ (request for quotation)`, navigate to the :guilabel:`Sales` app, and
click :guilabel:`New`. Fill out the inforamtion on the new quotation, and add the two products
stored in the two warehouses. 

Then, click the :guilabel:`Other Info` tab, and change the :guilabel:`Warehouse` to the virtual warehouse that was previously created. Once the quotation has been filled out, click :guilabel:`Confirm`.

Now that the quotation has been changed into a sales order, click the :guilabel:`Delivery` smart
button. From here, change the :guilabel:`Source Location` to the virtual warehouse created

> previously.

Reviewer 3: Now that the quotation has been changed to a sales order, click the :guilabel:`Delivery` smart
button. From here, change the :guilabel:`Source Location` to the virtual warehouse that was previously created.

.. note::
   The :guilabel:`Source Location` and warehouse under the :guilabel:`Other Info` tab must match

>    for the products included in the sales order to be pulled from different warehouses.

Reviewer 3: The :guilabel:`Source Location` and warehouse, under the :guilabel:`Other Info` tab, must match in order for the products included in the sales order to be pulled from different warehouses.

.. image:: stock_warehouses/stock-warehouses-delivery-order.png
   :align: center
   :alt: Delivery order with matching source and child locations.

Finally, if they are not set already, change the locations under the :guilabel:`From` column for
each product to the child locations previously tied to the virtual parent location. Once everything

> is set, :guilabel:`Validate` the delivery and invoice for the sales order.

Reviewer 3: Finally, if they aren't set already, change the locations under the :guilabel:`From` column for
each product to the "Child" locations that were previously tied to the virtual "Parent" location. 

Once everything is set, :guilabel:`Validate` the delivery, and invoice for the sales order.
